In [1]:
import os
import pandas as pd
import matplotlib as plt
%matplotlib inline

# Cleaning Manually Typed In Data by WD Students
For years, work duty students have courageously attempted to type in data from peabody cards into a spreadsheet. Their labor is central to our data-mining, hi-tech endeavor in trying to train a model to recognize letters *automatically*, like magic. Yet the data given to us by the Peabody remains muddy. Work Duty students are generally exhausted and disinterested.

## Discovering the CSV
We didn't know, initially, that this CSV existed. Below demonstrates our discovery of this CSV.

In [2]:
# The root path of the directory
# i.e., the folder containing all files related to this project. 
root_path = os.getcwd()
os.listdir(root_path)

['.DS_Store',
 '.git',
 '.ipynb_checkpoints',
 'Card Catalog Processing.ipynb',
 'Moe.ipynb',
 'Otsu_Thresholding_Cropping.ipynb',
 'pdftotiff.sh',
 'peabody_files',
 'README.md',
 'util.sh']

I know that **`peabody_files` contains all of the data supplied by the Peabody Museum.** For convenience, I store this path so it is easier to come back to.

In [3]:
data_path = os.path.join(root_path, 'peabody_files')
os.listdir(data_path)

['.DS_Store', 'Accession Files', 'Peabody_Extended_Fields_Acc.1-12.csv']

This is interesting. `Accession Files` seem to lead to more data, but there is another mysterious csv, the purposes of which cannot be discerned unless the data is examined.  

## Examining the CSV and Identifying Relevant Columns
### Examining CSV
The first step is to open the CSV.

In [4]:
csv_name = os.listdir(data_path)[2]
ext_df = pd.read_csv(os.path.join(data_path, csv_name)) #Short for extended fields
pd.set_option("display.max_columns", len(ext_df.keys())) #Display all columns
ext_df.sample(10)

ACCESSNO  ACQVALUE  APPNOTES  APPRAISOR CAPTION CAT      CATBY  \
640         2       0.0       NaN        NaN     NaN   O  Work Duty   
318         3       0.0       NaN        NaN     NaN   O  Work Duty   
456         3       0.0       NaN        NaN     NaN   O        NaN   
234         3       0.0       NaN        NaN     NaN   O  Work Duty   
120         3       0.0       NaN        NaN     NaN   O  Work Duty   
244         3       0.0       NaN        NaN     NaN   O  Work Duty   
621         2       0.0       NaN        NaN     NaN   O  Work Duty   
104         3       0.0       NaN        NaN     NaN   O  Work Duty   
775         2       0.0       NaN        NaN     NaN   O  Work Duty   
389         3       0.0       NaN        NaN     NaN   O  Work Duty   

        CATDATE      CATTYPE  CLASSES COLLDATE  COLLECTION COLLECTOR CONDDATE  \
640  03/29/2016  Archaeology      NaN      / /         NaN       NaN      / /   
318  05/03/2016  Archaeology      NaN      / /         NaN       NaN      / /   
456         / /  Archaeology      NaN      / /         NaN       NaN      / /   
234  05/03/2016  Archaeology      NaN      / /         NaN       NaN      / /   
120  04/20/2016  Archaeology      NaN      / /         NaN       NaN      / /   
244  05/04/2016  Archaeology      NaN      / /         NaN       NaN      / /   
621  03/28/2016  Archaeology      NaN      / /         NaN       NaN      / /   
104         / /  Archaeology      NaN      / /         NaN       NaN      / /   
775  03/30/2016  Archaeology      NaN      / /         NaN       NaN      / /   
389  05/10/2016  Archaeology      NaN      / /         NaN       NaN      / /   

    CONDEXAM CONDITION CONDNOTES  CREATOR  CREATOR2  CREATOR3  CREDIT  \
640      NaN       NaN       NaN      NaN       NaN       NaN     NaN   
318      NaN       NaN       NaN      NaN       NaN       NaN     NaN   
456      NaN       NaN       NaN      NaN       NaN       NaN     NaN   
234      NaN       NaN       NaN      NaN       NaN       NaN     NaN   
120      NaN       NaN       NaN      NaN       NaN       NaN     NaN   
244      NaN       NaN       NaN      NaN       NaN       NaN     NaN   
621      NaN       NaN       NaN      NaN       NaN       NaN     NaN   
104      NaN       NaN       NaN      NaN       NaN       NaN     NaN   
775      NaN       NaN       NaN      NaN       NaN       NaN     NaN   
389      NaN       NaN       NaN      NaN       NaN       NaN     NaN   

     CURVALMAX  CURVALUE  DATASET DATE  \
640        0.0       0.0      NaN  NaN   
318        0.0       0.0      NaN  NaN   
456        0.0       0.0      NaN  NaN   
234        0.0       0.0      NaN  NaN   
120        0.0       0.0      NaN  NaN   
244        0.0       0.0      NaN  NaN   
621        0.0       0.0      NaN  NaN   
104        0.0       0.0      NaN  NaN   
775        0.0       0.0      NaN  NaN   
389        0.0       0.0      NaN  NaN   

                                               DESCRIP  DIMTYPE  DISPVALUE  \
640          3 quartz arrowpoints. 1 shale arrowpoint.        2        NaN   
318       Large ovoid plummet. Original card: Plummet.        2        NaN   
456  Small polishing stone. Polishing surface on on...        2        NaN   
234  Long, narrow rhyolite point with excurvate edg...        2        NaN   
120  Piece of broken harpoon. Found in box of potsh...        2        NaN   
244  Small ovoid plummet blank with an elongated ne...        2        NaN   
621          Retouched chip arrowpoint of green shale.        2        NaN   
104  Beaver incisor with beveled distal end. Utiliz...        2        NaN   
775                         Broken quartz hammerstone.        2        NaN   
389  Cresent shaped slate fragment. Original Access...        2        NaN   

     EARLYDATE  EVENT  EXHIBITNO EXHLABEL1  EXHLABEL2 EXHLABEL3  EXHLABEL4  \
640          0    NaN          0       NaN        NaN       NaN        NaN   
318          0    NaN          0       NaN       

### Preliminary Eliminations of Columns

This seems like an incomprehensible table, so I would benefit from identifying which columns contain valuable information. My first impression from the data is that **there are a lot of NaNs**. The most useful columns to me would not be dominated by NaNs. 

In [5]:
ext_df.isnull().sum().describe(percentiles=[x*0.1 for x in range(1, 9)])

count     249.000000
mean      711.638554
std       504.345869
min         0.000000
10%         0.000000
20%         0.000000
30.0%       4.000000
40%      1054.000000
50%      1093.000000
60.0%    1094.000000
70%      1094.000000
80%      1094.000000
max      1094.000000
dtype: float64

From this I understand that half of the **60%** of the **249 columns** have 1054 or more NaNs, which brings down the potential relevant columns down to 100 columns.

In [6]:
# Store 100 columns with the least NaNs
fewer_nan_cols = ext_df.isnull().sum().sort_values().index[:100]

My second impression from this table is that **some columns contain the same values for all rows.** This would automatically disqualify a column from containing relevant data.

In [7]:
# Store columns whose values vary
unique_cols = []
for col in ext_df.columns:
    unique_vals = ext_df[col].unique()
    unique_count = len(unique_vals)
    if unique_count !=1:
        unique_cols.append(col)

From these two impressions, I can whittle down the number of columns to review from 249 to 60. These columns are shown below:

In [8]:
cols_to_review = [col for col in fewer_nan_cols if col in unique_cols]
ext_df[cols_to_review]

ACCESSNO WEBINCLUDE      UPDATEDBY     UPDATED UDF18 UDF13  STATUSDATE  \
0            3        .F.  Taylor, Marla  05/10/2016   / /     0  05/10/2016   
1            3        .F.  Taylor, Marla  05/10/2016   / /     0  05/10/2016   
2            3        .F.  Taylor, Marla  05/10/2016   / /     0  05/10/2016   
3            3        .F.  Taylor, Marla  05/10/2016   / /     0  05/10/2016   
4            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
5            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
6            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
7            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
8            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
9            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
10           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
11           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
12           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
13           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
14           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
15           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
16           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
17           3        .F.       Designer  01/12/2006   / /     0         / /   
18           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
19           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
20           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
21           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
22           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
23           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
24           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
25           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
26           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
27           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
28           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
29           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
...        ...        ...            ...         ...   ...   ...         ...   
1064         4        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
1065         5        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1066         5        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1067         5        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1068         6        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1069         6        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1070         6        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1071         6        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1072         6        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1073         7        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1074         7        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1075         7        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1076         7        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1077         7        .F.      Work duty  05/19/2016   / /     0  05/19/2016   
1078         7        .F.      Work duty  05/19/2016   / /     0  05/19/2016   
1079         7        .F.      Work duty  05/19/2016   / /     0  05/19/2016   
1080         7        .F.      Work duty  05/19/2016   / /     0  05/19/2016   
1081         7        .F.      Work duty  05/19/2016   / /     0  05/19/2016   
1082  

### Comparing CSV data with graphic data

With the table above, it is still difficult to identify which columns are helpful to building our OCR model and which are not. At this point, we can manually compare our cards and the dataset to see which columns contain data that correlates with the card text.

As Jupyter Notebooks are built on web standards, they allow the display of PDFs. For convenience I'll write quick functions:

In [9]:
from IPython.display import HTML
    
def card_path(acc_no, cat_no):
    return './peabody_files/Accession Files/%d_%04d.pdf' % (acc_no, cat_no)

def disp_card(acc_no, cat_no):
    return HTML('<iframe src=\'%s\' width=700 height=350></iframe>' % card_path(acc_no, cat_no)) 

def disp_cards(*args):
    '''
    Pass in multiple tuples. Tuples should be (acc_no, cat_no)
    '''
    html = ''
    for tup in args:
        frame = '<iframe src=\'%s\' width=700 height=350></iframe>' % card_path(*tup)
        html += frame
    return HTML(html)    

With the functions above, I can compare which columns have text that is also in the cards. Crucially, I notice that the OBJECTID column corresponds to 'Access number / category number', which helps me compare specific cards to rows.

In [10]:
ext_df[cols_to_review]

ACCESSNO WEBINCLUDE      UPDATEDBY     UPDATED UDF18 UDF13  STATUSDATE  \
0            3        .F.  Taylor, Marla  05/10/2016   / /     0  05/10/2016   
1            3        .F.  Taylor, Marla  05/10/2016   / /     0  05/10/2016   
2            3        .F.  Taylor, Marla  05/10/2016   / /     0  05/10/2016   
3            3        .F.  Taylor, Marla  05/10/2016   / /     0  05/10/2016   
4            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
5            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
6            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
7            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
8            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
9            3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
10           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
11           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
12           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
13           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
14           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
15           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
16           3        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
17           3        .F.       Designer  01/12/2006   / /     0         / /   
18           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
19           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
20           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
21           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
22           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
23           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
24           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
25           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
26           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
27           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
28           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
29           3        .F.      Work duty  04/19/2016   / /     0  04/19/2016   
...        ...        ...            ...         ...   ...   ...         ...   
1064         4        .F.      Work duty  05/17/2016   / /     0  05/17/2016   
1065         5        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1066         5        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1067         5        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1068         6        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1069         6        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1070         6        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1071         6        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1072         6        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1073         7        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1074         7        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1075         7        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1076         7        .F.  Taylor, Marla  05/19/2016   / /     0  05/18/2016   
1077         7        .F.      Work duty  05/19/2016   / /     0  05/19/2016   
1078         7        .F.      Work duty  05/19/2016   / /     0  05/19/2016   
1079         7        .F.      Work duty  05/19/2016   / /     0  05/19/2016   
1080         7        .F.      Work duty  05/19/2016   / /     0  05/19/2016   
1081         7        .F.      Work duty  05/19/2016   / /     0  05/19/2016   
1082  

In [11]:
disp_card(3, 1000)

Just for the card with Acc. No. 3 and Cat. No. 1000, I carefully determine that the columns ACCESSNO, OBJECTID, DESCRIP, PROVENANCE, SITENO, SITE, ORIGIN, OLDNO are relevant. I'll try this out for two more random cards

In [12]:
ext_df[cols_to_review].sample(2)

ACCESSNO WEBINCLUDE  UPDATEDBY     UPDATED UDF18 UDF13  STATUSDATE  \
102         3        .F.  Work duty  04/26/2016   / /     0  04/26/2016   
727         2        .F.  Work duty  04/04/2016   / /     0  04/04/2016   

          RECAS OBJECTID   MAINTDATE  DEPTHFT  WIDTHIN  WIDTHFT  WIDTH  \
102  Expedition    3/582         / /      0.0      0.0      0.0    0.0   
727  Expedition    2/193  03/29/2016      0.0      0.0      0.0    0.0   

     WEIGHTLB  WEIGHTIN  WEIGHT REPATNOTIC  DEPTH  LENGTHIN  LENGTH  \
102       0.0       0.0       0        / /    0.0       0.0     0.0   
727       0.0       0.0       0        / /    0.0       0.0     0.0   

    EXCAVADATE  DEPTHIN  LENGTHFT  IMAGENO CONDDATE     CATDATE      CATTYPE  \
102        / /      0.0       0.0        0      / /         / /  Archaeology   
727        / /      0.0       0.0        0      / /  04/04/2016  Archaeology   

    EXHSTART COLLDATE                                            DESCRIP  \
102      / /      / /  Base of large antler and cut several inched ab...   
727      / /      / /                            Broken shale arrowhead.   

             GPARENT            PARENT            OBJNAME  \
102  12: Archaeology  Antler, Modified        Antler Tool   
727  12: Archaeology    Stone, Chipped  Point, Projectile   

                                            PROVENANCE  SITENO  \
102  Section number - 4E Trench Number - IC.3 Excav...  42.001   
727  East end of Squibnocket Pond, Martha's Vineyar...   M49/2   

                      SITE         MATERIAL   STATUSBY   STATUS      CATBY  \
102        Nevin Shellheap  Faunal Material  Work Duty  Entered  Work Duty   
727  Hornblower Shell Heap            Stone  Work Duty  Entered  Work Duty   

                                       ORIGIN OLDNO  \
102  United States | ME | Hancock | Blue Hill   181   
727                                       NaN   NaN   

                                                 NOTES  COUNT OTHERNO HOMELOC  \
102  Douglas Byers (Principal Investigator, 00/00/3...    1.0     NaN    RSPM   
727                                                NaN    NaN   SA/61     NaN   

    LOCFIELD1 STRATUM                  LEVEL                          UDF3  \
102       NBS      26  In humus below shells  Late Archaic Moorehead Phase   
727       NaN     NaN                    NaN                           NaN   

       UNIT          PEOPLE      EXCAVATEBY  LOCFIELD2  LOCFIELD3 REPATTYPE  \
102  IC3/4E  Byers, Douglas  Byers, Douglas        NaN        NaN       NaN   
727     NaN             NaN             NaN        NaN        NaN       NaN   

    PUBNOTES UDF4 UDF21  
102      NaN  NaN   NaN  
727      NaN  NaN   NaN

In [13]:
disp_card(3, 911)
disp_card(2, 283)

From the three above
- **ACCESSNO** has the Acc. No. verbatim
- **OBJECTID** has 'Acc. No/Cat. No.' verbatim
- **DESCRIP** matches Name (most of the time. Not during abbreviations, like 2/283)
- **OLDNO** and **OTHERNO** matches Orig.No. It's likely that if one is NaN, the other is filled.
- **PROVENANCE** matches card's 'Locality,' 'Situation,' and 'Remarks' but formatted weirdly.
- **SITENO** seems to correlate with cards' 'Site No.' category, but not sure how.
- **SITE** matches card's 'Site' category, but text is formatted differently. (Nevin Shellheap and Nevin Shell Heap)
- **ORIGIN** matches Locality when it's there.
- **NOTES** matches remarks sometimes 

## Cleaning the data
Now that the relevant columns have been identified, it's time to clean the data. **The objective of this is to process the text in the csv to match the text on the cards as closely as possible.**

### Observations

In [14]:
cols_to_review = ['ACCESSNO','OBJECTID','DESCRIP','OLDNO','OTHERNO','PROVENANCE','SITENO','SITE','ORIGIN','NOTES']
ext_df[cols_to_review]

ACCESSNO OBJECTID                                            DESCRIP  \
0            3   3/1000                                        Pot sherds.   
1            3   3/1001                    Fragment of chipped arrowpoint.   
2            3   3/1002                                        Pot sherds.   
3            3   3/1003                            Beaver tooth implement.   
4            3   3/1005  Barbed harpoon fragment. Appears to be unilate...   
5            3   3/1006                     2 fragments stone arrow points   
6            3   3/1007                              1 chipped stone knife   
7            3   3/1008                              Butt of adze? Surface   
8            3   3/1009                       Edge of polished slate tool.   
9            3   3/1010                                     Large plummet.   
10           3   3/1011               Point of large broken chipped knife.   
11           3   3/1012                              Felsite (?) core. Sod   
12           3   3/1013  Small plummet, sandstone. In mixed shell and h...   
13           3   3/1014                           Point of bone pike head.   
14           3   3/1015                   Fragment of worked beaver tooth.   
15           3   3/1016                              Crudely chipped knife   
16           3   3/1017  Distal end of large bone perferator with cut m...   
17           3   3/1152                     Large flake used as a chopper.   
18           3    3/491  Chipped triangular biface preform. Ovate, stra...   
19           3    3/492                                 Stemmed arrowpoint   
20           3    3/493  Two chipped stone edge tools and one groundsto...   
21           3    3/494   One biface preform. Rough chipped felsite block.   
22           3    3/495  Chipped crescent-shaped chopper. Chipped semi-...   
23           3    3/496                                       Pottery rim.   
24           3    3/497                                  Small pot sherds.   
25           3    3/500                                       Broken celt.   
26           3    3/501  Grooved cap of large plummet. Broken knobbed p...   
27           3    3/502                             Broken felsite biface.   
28           3    3/503  Long, narrow rhyolite point with broken tip, e...   
29           3    3/504                             Crude chipped preform.   
...        ...      ...                                                ...   
1064         4   4/1061                                   Quartzite knife.   
1065         5   5/1062                       Base of quartzite spearhead.   
1066         5   5/1063                           Base of quartzite knife.   
1067         5   5/1064                                    Notched sinker.   
1068         6   6/1065                                    Quartzite chip.   
1069         6   6/1067                                 Base of arrowhead.   
1070         6   6/1068                                  Quartz spearhead.   
1071         6   6/1069                       Triangular quartz arrowhead.   
1072         6   6/1070                                  Coiled pot sherd.   
1073         7   7/1071                                          Bone awl.   
1074         7   7/1072                                 Piece of cut bone.   
1075         7   7/1073                                 Piece of cut bone.   
1076         7   7/1074               Single-barbed bone fish spear point.   
1077         7   7/1075                                          Bone awl.   
1078         7   7/1076               Single-barbed bone fish spear point.   
1079         7   7/1077                         Fragment of bird bone awl.   
1080         7   7/1078               Single-barbed bone fish spear point.   
1081         7   7/1079               Single-barbed bone fish spear point.   
1082         7   7/1080                                       Poll of axe.   
1083         7   7/1081     

In [15]:
df = ext_df[cols_to_review]
df.isnull().sum()

ACCESSNO        0
OBJECTID        0
DESCRIP         1
OLDNO         541
OTHERNO       585
PROVENANCE      2
SITENO          7
SITE            8
ORIGIN        406
NOTES         562
dtype: int64

In [16]:
for col in df.columns:
    print(col, len(df[col].unique()), sep=' | ')

ACCESSNO | 12
OBJECTID | 1094
DESCRIP | 877
OLDNO | 518
OTHERNO | 363
PROVENANCE | 761
SITENO | 13
SITE | 14
ORIGIN | 8
NOTES | 38


Additional takeaways:
- `ACCESSNO` and `OBJECTID` are the most complete and reliable dataset.
- `DESCRIP` is a close second in completeness, but has so much variety that it would be difficult to eliminate differences from cards.
- `SITENO` and `SITE` have few NaNs and few variety, making them nearly as ideal as `ACCESSNO` and `OBJECTID`.
- That `OLDNO` and `OTHERNO` have high number of NaNs suggests that when one is NaN, the other is filled.

**I will start by focusing on Acc. No., Cat. No, and Site. **

### Acc. No. and Cat. No.
Goal: A new column `CATEGORYNO` holds result of text-processed `OBJECTID` showing only category number.

#### Checking Assumptions Before Text processing
My assumption is that all the values in `OBJECTID` follow the format ACCNO/CATNO. I want to check if this is true.

In [17]:
# Check if all values in OBJECTID column follow the format number/number
import re
follows_format = df.OBJECTID.apply(lambda s: bool(re.search('\d+\/\d+', s)))
follows_format.all()

False

Not all the values follow the format. What do the exceptions look like?

In [18]:
i = follows_format[follows_format == False].index
df.iloc[i]

ACCESSNO OBJECTID                                            DESCRIP  \
515          7   Jul-78                                                NaN   
520          1    1-Jan                                  Butt of arrowhead   
521          1    2-Jan                               Butt of quartz knife   
522          1    3-Jan                                 Crude quartz point   
523          1   14-Jan                       Point of large quartz blade.   
524          1    4-Jan                             Crude quartzite spear.   
525          1   15-Jan   Quartz arrow head, thick butt point chipped off.   
526          1    5-Jan                                     Crude chopper.   
527          1    6-Jan                  Crude rectangular quartz scraper.   
528          1   16-Jan        Red quartzite (?) arrowhead, straight butt.   
529          1    7-Jan                            Crude quartz arrowhead.   
530          1   17-Jan                    Crude arrow point, not stemmed.   
531          1    8-Jan                                Crude broken point.   
532          1   18-Jan                  Crude knife (?), quartz, stemmed.   
533          1    9-Jan                                     Rubbing stone.   
534          1   10-Jan                       Edge of crude chipped blade.   
535          1   11-Jan                      Rounded butt of quartz knife.   
536          1   19-Jan                                Quartz arrow point.   
537          1   12-Jan                          Point of crude arrowhead.   
538          1   20-Jan                         Butt of large crude knife.   
539          1   13-Jan                       Point of large quartz blade.   
540          1   21-Jan                           Broken quartz arrowhead.   
541          1   23-Jan                                   Crude arrowhead.   
542          1   Jan-35                                   One quartz arrow   
543          1   22-Jan                           Broken quartz arrowhead.   
544          1   24-Jan                                 Butt of arrowhead.   
545          1   Jan-36                                       Quartz arrow   
546          1   25-Jan                                       Quartz Knife   
547          1   26-Jan                                       Quartz Knife   
548          1   Jan-37                                       Broken arrow   
...        ...      ...                                                ...   
590          1   Jan-61                                 Piece of bone awl.   
591          1   Jan-62                                        Cut antler.   
592          2   Feb-85                        2 small quartz arrowpoints.   
593          1   Jan-63                           Scraps of crude pottery.   
594          1   Jan-64                 Two pieces of undecorated pottery.   
595          1   Jan-65                  First shell layer above scallops.   
596          1   Jan-66                             Small piece of pottery   
597          2   Feb-76                            Small quartz arrowpoint   
598          2   Feb-77                               2 Quartz Hammerstone   
600          2   Feb-78                               4 Quartz Arrowpoints   
601          2   Feb-79                        5 Fragments of quartz tools   
603          2   Feb-80                                     Quartz scraper   
604          2   Feb-81              1 arrowpoint and 1 butt of arrowpoint   
606          2   Feb-82                            Fragment of shale knife   
608          2   Feb-83                           Piece of worked hematite   
610          2   Feb-84                                  Quartz turtleback   
612          2   Feb-86                         2 fragments of crude tools   
614          2   Feb-87  Part of large broken knife 1 quartz arrowpoint...   
618          2   Feb-89                           Broken quartz arrowpoint   
619          2   Feb-90     

Hypothesis: If values in `OBJECTID` don't follow the Number/Number format, they follow the Month-Number or Number-Month format. 
Let's test this:

In [19]:
months_abbv = ['Jan','Feb','March','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec']
re_months = '(%s)' % ('|'.join(months_abbv))
t = df.iloc[i].OBJECTID.apply(lambda s: bool(re.match('(%s\-\d+)|(\d+\-%s)'%(re_months, re_months), s)))
t[t==False]

1013    False
Name: OBJECTID, dtype: bool

In [20]:
df.iloc[1013]

ACCESSNO                                                      3
OBJECTID                                                      3
DESCRIP                                           Jaw and teeth
OLDNO                                                       158
OTHERNO                                                     NaN
PROVENANCE    Locality: Bluehill Falls, Me. Section number -...
SITENO                                                   42.001
SITE                                            Nevin Shellheap
ORIGIN                 United States | ME | Hancock | Blue Hill
NOTES         Douglas Byers (Principal Investigator, 00/00/3...
Name: 1013, dtype: object

In [21]:
len([n for n in os.listdir('./peabody_files/Accession Files/') if n[0]=='3'])

526

For some reason, this row at index **1013** does mention Cat. No. in `OBJECTID`. The `ACCESSNO` of this row is 3, but there are **526** possible cards whose `ACCESSNO` is 3. Thus, **it is unfeasible to determine which card this row corresponds to.** 

In [22]:
df = df.drop(1013) #Bye!

There are three patterns for the `OBJECTID` column:
- ACCNO/CATNO
- MONTH-CATNO
- CATNO-MONTH

We want to convert all of these into *just* CATNO. 

In [23]:
def get_catno(string):
    if '/' in string:
        return int(re.search('(?<=\/)\d+',string).group(0))
    return int(re.search('\d+', string).group(0))

df['CATEGORYNO'] = df['OBJECTID'].apply(get_catno)
df.sample(5)

ACCESSNO OBJECTID                                            DESCRIP  \
401         3    3/926  Broken arrow point. Lanceolate, straight base....   
268         3    3/762  Bone harpoon with al large drilled hole and a ...   
19          3    3/492                                 Stemmed arrowpoint   
8           3   3/1009                       Edge of polished slate tool.   
102         3    3/582  Base of large antler and cut several inched ab...   

    OLDNO OTHERNO                                         PROVENANCE  SITENO  \
401   203     NaN  Provenience Description - Above plowline below...  42.001   
268   277     NaN  Section number - 7E Trench Number - IC.2 Excav...  42.001   
19      2     NaN  Section number - 2 Trench Number - ID.2 Locali...  42.001   
8     13C     NaN                       Bluehill Falls, Me. Surface.  42.001   
102   181     NaN  Section number - 4E Trench Number - IC.3 Excav...  42.001   

                SITE                                    ORIGIN  \
401  Nevin Shellheap  United States | ME | Hancock | Blue Hill   
268  Nevin Shellheap  United States | ME | Hancock | Blue Hill   
19   Nevin Shellheap  United States | ME | Hancock | Blue Hill   
8    Nevin Shellheap  United States | ME | Hancock | Blue Hill   
102  Nevin Shellheap  United States | ME | Hancock | Blue Hill   

                                                 NOTES  CATEGORYNO  
401                   July 2013 PastPerfect Conversion         926  
268  Douglas Byers (Principal Investigator, 00/00/3...         762  
19                    July 2013 PastPerfect Conversion         492  
8                     July 2013 PastPerfect Conversion        1009  
102  Douglas Byers (Principal Investigator, 00/00/3...         582

Looks good. To be very meticulous, I also want to check another assumption: that ACCNO from `OBJECTID` match with `ACCESSNO` column. 

In [24]:
def get_accno(string):
    if '/' in string:
        return int(re.match('\d+',string).group(0))
    month = re.search('[^-\d]+', string).group(0)
    return int(months_abbv.index(month)+1)

df['ACCESSNO_OBJ'] = df['OBJECTID'].apply(get_accno)
df.query('ACCESSNO_OBJ != ACCESSNO')

Empty DataFrame
Columns: [ACCESSNO, OBJECTID, DESCRIP, OLDNO, OTHERNO, PROVENANCE, SITENO, SITE, ORIGIN, NOTES, CATEGORYNO, ACCESSNO_OBJ]
Index: []

Another assumption is that all of the rows are unique to a ACCNO-CATNO pair.

In [25]:
df.duplicated(subset=['ACCESSNO','CATEGORYNO']).any()

False

Another assumption is that all of the rows have a corresponding image. Is this true?

In [26]:
def image_exists(x):
    return os.path.exists(card_path(x['ACCESSNO'], x['CATEGORYNO']))

t = df.apply(image_exists, axis=1)
len(t[t==False])

23

**23 rows** do not have an image of a card that corresponds to their data. These are useless to our purpose of training an OCR, so we drop those too.

In [27]:
i = t[t==False].index
df.drop(i, inplace=True)

Now, it is 100% definite that all of the rows in the `df` dataset have the correct values in the `ACCESSNO` and `CAtEGORYNO` columns, and that all the rows are unique to one card that exists.

### Reevaluating Which Other Fields to Pay Attention To
In the section above, we dropped a number of rows that did not fulfill the criteria: 
-  Could not be identified by Category and Accession number
- Did not correspond to an image. 

Because of this, I want to re-evaluate which columns I want to pay attention to.

In [28]:
df.isnull().sum()

ACCESSNO          0
OBJECTID          0
DESCRIP           0
OLDNO           519
OTHERNO         562
PROVENANCE        0
SITENO            1
SITE              1
ORIGIN          389
NOTES           546
CATEGORYNO        0
ACCESSNO_OBJ      0
dtype: int64

In [29]:
for col in df.columns:
    print(col, len(df[col].unique()), sep=' | ')

ACCESSNO | 7
OBJECTID | 1070
DESCRIP | 861
OLDNO | 517
OTHERNO | 362
PROVENANCE | 749
SITENO | 8
SITE | 10
ORIGIN | 6
NOTES | 35
CATEGORYNO | 1070
ACCESSNO_OBJ | 7


The columns **SITE** and **SITENO** remain interesting, due to its few variety and **one** NaN. I will start there.

### SITE and SITENO

In [30]:
pd.set_option('max_colwidth',200)
df[['ACCESSNO','CATEGORYNO','SITE', 'SITENO']].sample(10)

ACCESSNO  CATEGORYNO                   SITE     SITENO
144          3         627        Nevin Shellheap     42.001
586          2          75  Hornblower Shell Heap      M49/2
276          3         771        Nevin Shellheap     42.001
122          3         605        Nevin Shellheap     42.001
958          2         458  Hornblower Shell Heap      M49/2
449          3         976        Nevin Shellheap     42.001
48           3         523        Nevin Shellheap     42.001
1070         6        1068      Mooseup Camp Site  Site 0821
914          2         381  Hornblower Shell Heap      M49/2
990          2         479  Hornblower Shell Heap      M49/2

Questions:
- Where is the NaN? (1 NaN in SITE column, 1 NaN in SITENO column)
- What are the unique values of SITE? What are the unique values of SITENO?
- Do SITE and SITENO always have the same pair?
- Do the cards correspond exactly to text on SITE or SITENO?

#### NaN

In [31]:
# Display the row with NaN in SITE column.
df[df['SITE'].isnull()]

ACCESSNO OBJECTID  \
334         3    3/857   

                                                                                DESCRIP  \
334  Bit end of humped-back gouge. Original Accession Card: Battered hump-backed gouge.   

    OLDNO OTHERNO  \
334   462     NaN   

                                                                                                                                                   PROVENANCE  \
334  Section number - 8E Trench Number - IC.1 Excav Date - 00/00/36 Locality - Bluehill Falls, Me. Situation - Just under shells in mixed shells and pebbles.   

    SITENO SITE ORIGIN  \
334    NaN  NaN    NaN   

                                                                                 NOTES  \
334  Douglas Byers (Principal Investigator, 00/00/36) July 2013 PastPerfect Conversion   

     CATEGORYNO  ACCESSNO_OBJ  
334         857             3

As `SITENO` is NaN too, this is the only row in `df` with NaN for Site and Site. No. Why did this happen? Was the information not displayed on the card?

In [32]:
disp_card(3, 857)

The Work Duty student was probably exhausted that way. I will manually fill in this Missing Value, but only because there is only one, and does not take too much time at all.

In [33]:
df.set_value(334, 'SITE', 'Nevin Shell Heap.')
df.set_value(334, 'SITENO', 'Me.152/1')

ACCESSNO OBJECTID  \
0            3   3/1000   
1            3   3/1001   
2            3   3/1002   
3            3   3/1003   
4            3   3/1005   
5            3   3/1006   
6            3   3/1007   
7            3   3/1008   
8            3   3/1009   
9            3   3/1010   
10           3   3/1011   
11           3   3/1012   
12           3   3/1013   
13           3   3/1014   
14           3   3/1015   
15           3   3/1016   
16           3   3/1017   
18           3    3/491   
19           3    3/492   
20           3    3/493   
21           3    3/494   
22           3    3/495   
23           3    3/496   
24           3    3/497   
25           3    3/500   
26           3    3/501   
27           3    3/502   
28           3    3/503   
29           3    3/504   
30           3    3/505   
...        ...      ...   
1048         4   4/1045   
1049         4   4/1046   
1050         4   4/1047   
1051         4   4/1048   
1052         4   4/1049   
1053         4   4/1050   
1054         4   4/1051   
1055         4   4/1052   
1056         4   4/1053   
1057         4   4/1054   
1058         4   4/1055   
1059         4   4/1056   
1060         4   4/1057   
1061         4   4/1058   
1062         4   4/1059   
1063         4   4/1060   
1064         4   4/1061   
1065         5   5/1062   
1066         5   5/1063   
1067         5   5/1064   
1068         6   6/1065   
1069         6   6/1067   
1070         6   6/1068   
1071         6   6/1069   
1072         6   6/1070   
1073         7   7/1071   
1074         7   7/1072   
1075         7   7/1073   
1076         7   7/1074   
1086         6   6/1066   

                                                                                                                                     DESCRIP  \
0                                                                                                                                Pot sherds.   
1                                                                                                            Fragment of chipped arrowpoint.   
2                                                                                                                                Pot sherds.   
3                                                                                                                    Beaver tooth implement.   
4                                                            Barbed harpoon fragment. Appears to be unilateral. Neither tip nor butt extant.   
5                                                                                                             2 fragments stone arrow points   
6                                                                                                                      1 chipped stone knife   
7                                                                                                                      Butt of adze? Surface   
8                                                                                                               Edge of polished slate tool.   
9                                                                                                                             Large plummet.   
10                                                                                                      Point of large broken chipped knife.   
11                                                                                                                     Felsite (?) core. Sod   
12                                                         Small plummet, sandstone. In mixed shell and humus near edge of bank above beach.   
13                                                                                                                  Point of bone pike head.   
14                                                                                                          Fragment of worked beaver tooth.   
15                         

### Unique Values

In [34]:
print('Site', ', '.join(df['SITE'].unique()), sep=' : ')
print('Site No.', ', '.join(df['SITENO'].unique()), sep=' : ')

Site : Nevin Shellheap, Nevin Shell Heap., Squibnocket Cliff Shellheap, Hornblower Shell Heap, E.D Prey c7/1, Pomfret Landing, Pomfret landing, Mooseup Camp Site, Mooseup camp Site, Shellheap #1, Plum Island
Site No. : 42.001, Me.152/1, Site 0213, M49/2, Site 1255, Site 0820, Site 0821, Site 0843


The values that I inserted before for `df.loc[334]` show up as unique values. This is not good, because I added another category altogether; I'll replace with those that were existing before. For `SITENO`, it seems from glancing through the table that Nevin Shellheap rows also have 42.001 as their `SITENO`, so I'll use that.

In addition, there are two unique values for `Pompfret landing` and `Pompfret Landing`, in addition to `Mooseup Camp Site` and `Mooseup camp Site`, so I will fix this, too.

In [35]:
# Fixing my manually inputed values for row with NaN
df.set_value(334, 'SITE', 'Nevin Shellheap')
df.set_value(334, 'SITENO', '42.001')

# Eliminating duplicate site names with different capitalizations
df.replace('Pomfret landing', 'Pomfret Landing', inplace=True)
df.replace('Mooseup camp Site', 'Mooseup Camp Site', inplace=True)

# Checking
print('Site', ', '.join(df['SITE'].unique()), sep=' : ')
print('Site No.', ', '.join(df['SITENO'].unique()), sep=' : ')

Site : Nevin Shellheap, Squibnocket Cliff Shellheap, Hornblower Shell Heap, E.D Prey c7/1, Pomfret Landing, Mooseup Camp Site, Shellheap #1, Plum Island
Site No. : 42.001, Site 0213, M49/2, Site 1255, Site 0820, Site 0821, Site 0843


### SITE and SITENO correlation

With only unique values, there are as many unique values of `SITE` as `SITENO`, strengthening the theory that every `SITE` has a `SITENO` corresponding to it. 

In [36]:
print(len(df['SITE'].unique()), len(df['SITENO'].unique()))

7 7


The theory would be verified if there are as many combinations of `SITE` and `SITENO` as the unique values of `SITE` and `SITENO`. In short, if there are **7** combinations of SITE and SITENO.

In [37]:
def g(x):
    return x['SITE'] + x['SITENO']

t = df.apply(g, axis=1)
len(t.unique())

7

Now, it's confirmed that **every site has a unique site number.** 

### Matching text typed on cards
Earlier on I noticed that, while the cards indicated Nevin Shell Heap, the data showed Nevin Shellheap. In addition, when I typed in data manually for the row with NaN, I noticed that the `SITENO` column in the data did not correspond to card's text. As the objective of this Jupyter Notebook is to try to have a dataset matching the card's text **as closely as possible**, I need to change the styling of the dataset to resemble the text.

In [38]:
def disp_cards_at_indexes(*args):
    '''
    Pass in labels of index axes.
    '''
    k = []
    for i in args:
        acc, cat = df.loc[i,'ACCESSNO'], df.loc[i, 'CATEGORYNO']
        k.append((acc, cat))
        
    return disp_cards(*k)

def disp_cards_at_site(site, n=6):
    x = df[df['SITE'] == site]
    if n > len(x):
        n = len(x)
    i = x.sample(n).index.tolist()
    return disp_cards_at_indexes(*i)

#### Nevin Shell Heap.

In [39]:
sites = df['SITE'].unique()
snos = df['SITENO'].unique()
disp_cards_at_site(sites[0])

In [40]:
df.replace(sites[0], 'Nevin Shell Heap.', inplace=True)
df.replace(snos[0], 'Me.152/1', inplace=True)

#### Squibnocket Cliff.

In [41]:
disp_cards_at_site(sites[1])

In [42]:
df.replace(sites[1], 'Squibnocket Cliff.', inplace=True)
df.replace(snos[1], '', inplace=True)

#### Hornblower Shell Heap.

In [43]:
disp_cards_at_site(sites[2])

In [44]:
df.replace(sites[2], 'Hornblower Shell Heap.', inplace=True)
df.replace(snos[2], '', inplace=True)

#### E.D. Prey.

In [45]:
disp_cards_at_site(sites[3])

In [46]:
df.replace(sites[3], 'E.D. Prey.', inplace=True)
df.replace(snos[3], 'C 7/1', inplace=True)

#### Pomfret Landing.

In [47]:
disp_cards_at_site(sites[4])

In [48]:
df.replace(sites[4], 'Pomfret Landing.', inplace=True)
df.replace(snos[4], 'C 7/3', inplace=True)

#### Moosup Camp Site.

In [49]:
disp_cards_at_site(sites[5])

In [50]:
df.replace(snos[5], 'C 14/1', inplace=True)

There is a problem with Moosup Camp Site: some cards have a capitalized s, while others have a capitalized S. If there are not many cards with Mooseup Camp Site, then I don't mind manually changing them. 

In [51]:
len(df[df['SITE']==sites[5]])

6

There are only 6 cards, so the cards displayed above are all of the cards I need to see. Cards with Cat. No's 1069 and 1070 have lower 's', and the rest have capitalized 'S's. According to the table below, those rows have indexes 1071 and 1072.

In [52]:
df[df['SITE']==sites[5]]

ACCESSNO OBJECTID                       DESCRIP OLDNO OTHERNO  \
1068         6   6/1065               Quartzite chip.   NaN     NaN   
1069         6   6/1067            Base of arrowhead.   NaN     NaN   
1070         6   6/1068             Quartz spearhead.   NaN     NaN   
1071         6   6/1069  Triangular quartz arrowhead.   NaN     NaN   
1072         6   6/1070             Coiled pot sherd.   NaN     NaN   
1086         6   6/1066       Base of quartzite tool.   NaN     NaN   

                                                                                           PROVENANCE  \
1068  Situation - Unknown. Locality - Plainfiled township, Conn. 5/8 mi. N.E. of head of Moosup Pond.   
1069      Sitaution:Unknown. Locality: Plainfield township, Conn. 5/8 mi. N.E.of head of Moosup Pond.   
1070    Situation: Unknown. Locality: Plainfield township, Conn. 5/8 mi. N.E. of head of Moosup Pond.   
1071    Situation: Unknown. Locality: Plainfield township, Conn. 5/8 mi. N.E. of head of Moosup Pond.   
1072    Situation: Unknown. Locality: Plainfield township, Conn. 5/8 mi. N.E. of head of Moosup Pond.   
1086  Situation - Unknown. Locality - Plainfiled township, Conn. 5/8 mi. N.E. of head of Moosup Pond.   

      SITENO               SITE ORIGIN NOTES  CATEGORYNO  ACCESSNO_OBJ  
1068  C 14/1  Mooseup Camp Site    NaN   NaN        1065             6  
1069  C 14/1  Mooseup Camp Site    NaN   NaN        1067             6  
1070  C 14/1  Mooseup Camp Site    NaN   NaN        1068             6  
1071  C 14/1  Mooseup Camp Site    NaN   NaN        1069             6  
1072  C 14/1  Mooseup Camp Site    NaN   NaN        1070             6  
1086  C 14/1  Mooseup Camp Site    NaN   NaN        1066             6

In [53]:
df.replace(sites[5], 'Moosup Camp Site.', inplace=True)
for i in [1071, 1072]:
    df.set_value(i, 'SITE', 'Moosup Camp site.')

#### Shell Heap \#1.

In [54]:
disp_cards_at_site(sites[6])

In [55]:
df.replace(sites[6], 'Shell Heap #1.', inplace=True)
df.replace(snos[6], '', inplace=True)

In [56]:
df.SITE.unique()

array(['Nevin Shell Heap.', 'Squibnocket Cliff.', 'Hornblower Shell Heap.',
       'E.D. Prey.', 'Pomfret Landing.', 'Moosup Camp Site.',
       'Moosup Camp site.', 'Shell Heap #1.'], dtype=object)

# Wrap Up

In [69]:
df.set_index(df.apply(lambda x: '%d_%04d'%(x['ACCESSNO'], x['CATEGORYNO']),axis=1), inplace=True)
final_df = df[['ACCESSNO','CATEGORYNO','SITE','SITENO']]
final_df.sample(6)

ACCESSNO  CATEGORYNO                    SITE    SITENO
1_0034         1          34      Squibnocket Cliff.          
3_0876         3         876       Nevin Shell Heap.  Me.152/1
3_0788         3         788       Nevin Shell Heap.  Me.152/1
3_0748         3         748       Nevin Shell Heap.  Me.152/1
2_0155         2         155  Hornblower Shell Heap.          
3_0605         3         605       Nevin Shell Heap.  Me.152/1

In [70]:
final_df.to_csv('transcription.csv')